### DISCUSSION TITLE
Can I detect shift between two images and output the result to a text file?

### DISCUSSION THREAD

**Discussion by jumpjack** [ImageMagick/ImageMagick#4021] [Digital Image Processing]
As per title, can I use Imagemagick to detect shift between two images and output the result to a text file?
And what about rotation?

**Comment by fmw42** [comment]
The compare function allows you to detect the shift of a smaller image relative to a larger image. Resulting offsets can be redirected to a text file.

Rotations are  not found with compare.  What platform/OS are you using?  

See 

https://legacy.imagemagick.org/Usage/compare/
https://imagemagick.org/script/compare.php
https://github.com/ImageMagick/ImageMagick/discussions/3950

**Reply by fmw42** [reply]
There are some techniques that ImageMagick can do when scripted that would get the rotation and scale between two images. But scripting is OS dependent. As asked above, what is your platform and ImageMagick version?

**Comment by fmw42** [comment]
Can you provide an example pair of images for which you want to get the offset and rotation?

**Comment by jumpjack** [comment]
I am using Windows 10.
What I am trying to do is to build a 360 panorama video stabilizer; I read that I could do it by identifying x/y shift of a small area of a big equirectangular frame w.r.t. previous frame, and using these data as an input for ffmpeg to specify the needed pitch/roll correction to apply to output, to compensate for shifting.

Frame 1:
![image](https://user-images.githubusercontent.com/1620953/127896199-662ae5ba-0d42-480a-afa2-9882ac43d957.png)

Frame 2:
![image](https://user-images.githubusercontent.com/1620953/127896349-0a5c5fd6-9249-487c-939e-2ff088fda7d5.png)

Frame 1 annotated:
![image](https://user-images.githubusercontent.com/1620953/127897383-cee695a4-1cc0-4bd7-b3e6-aa69c5708b01.png)

Rotation is a secondary requirement.

**Comment by fmw42** [comment]
I renamed your image to large.png and cropped a small area of size 50x50 at offsets=+190+160 as:

![small](https://user-images.githubusercontent.com/13458218/127924892-216a5190-b462-4c3e-b10d-180515e75c18.png)

I then ran compare as:

`magick compare -metric NCC -subimage-search large.png small.png corr.png`
0.9996 @ 190,160

So the match score was 0.9996 out of 1 where 1 is a perfect match. The location of the match was at 190,160 exactly where cropped.

The correlation image was:

![corr-1](https://user-images.githubusercontent.com/13458218/127925067-afd59318-085b-47e0-bddf-722c86998eda.png)

The brightest spot is where it matched.

To send to a text file, you would redirect as

`magick compare -metric NCC -subimage-search large.png small.png null: 2> textfile.txt`

**Reply by jumpjack** [reply]
As your commandline was taking minutes and 100% processing power of a Core <s>i7</s>  i5  6-core processor using IM7, I tried a simpler processing:
- crop a 100x100 area from frame 1
- crop a 50x50 area from frame 2
- look for crop2 into crop 1

It worked... but it took anyway 20 seconds! Is that normal, or can I tune something?

Crop 1:
![frame-01-crop](https://user-images.githubusercontent.com/1620953/128040975-134c2240-bdf6-49d5-8f52-5ae77202fa65.png)

Crop 2:
![frame-02-crop](https://user-images.githubusercontent.com/1620953/128040998-75655d2b-8a33-4365-9b44-b71607320ba0.png)

**Reply by fmw42** [reply]
My command should have been very quick if you are using a recent version of IM 7 with HDRI and have FFTW delegate installed for Fourier Transforms (FFT). Cropping should have been very quick. 20 sec is way to long.  What is your IM 7 version? How much RAM do you have?

`time magick compare -metric NCC -subimage-search large.png small.png corr.png`
0.956972 @ 190,162

real	0m1.756s
user	0m5.215s
sys	0m0.582s

So took less than 2 sec on my Mac OSX with 2 cores.  For this speed on compare, you need IM 7.1.0-3 or higher or it will be slow as it won't use FFT processing

For the cropping:

`time magick large.png -crop 100x100+100+100 +repage cropped.png`

real	0m0.278s
user	0m0.105s
sys	0m0.113s

This took less than 0.3 sec.

**Reply by jumpjack** [reply]
I'm on 7.0.8-53 q16 X64, let me upgrade and try again (but I have 20 GB of RAM and i5, not i7).

Anyway metric RMSE appears much faster than NCC.

**Reply by fmw42** [reply]
RMSE is faster than NCC before IM 7.1.0.3.  For IM 7.1.0.3 and later if you have HDRI and FFTW, then NCC is 1000 times faster than before due to new implementation as per https://github.com/ImageMagick/ImageMagick/discussions/3950

**Reply by jumpjack** [reply]
```
@echo off
set TA=%time% & echo Started at: %time%
magick compare -metric NCC -subimage-search frame-01-crop.png frame-02-crop.png resultNCC.png
set TB=%time% & echo. & echo Finished at: %time%
```

```
Started at: 19.03.07,69
0.937629 @ 32,32
Finished at: 19.03.30,81
```

![resultNCC-0](https://user-images.githubusercontent.com/1620953/128056974-fc036ef3-2866-4a68-bdca-d12ff77267a0.png)

23 seconds

----------

```
@echo off
set TA=%time% & echo Started at: %time%
magick compare -metric RMSE -subimage-search frame-01-crop.png frame-02-crop.png resultRMSE.png
set TB=%time% & echo. & echo Finished at: %time%
```

```
Started at: 19.04.55,43
5125.85 (0.0782154) @ 32,32
Finished at: 19.04.55,66

```

![resultRMSE-0](https://user-images.githubusercontent.com/1620953/128057004-c0970fee-7507-4b34-9850-9692a789366b.png)


Less than 1 second, same graphical result, don't understand text result.

**Reply by fmw42** [reply]
You have the start and finish times. The values in the middle (0.937629 @ 32,32) are the correlation value and the x,y location of the match i.e. the top left corner of the red rectangle in the output image you show.

**Comment by jumpjack** [comment]
Looking for cropping in whole image, using RMSE:

```
Started at: 19.12.45,30
5125.93 (0.0782167) @ 189,174
Finished at: 19.12.48,24
```

![resultRMSE-full-0](https://user-images.githubusercontent.com/1620953/128057809-8b52d10c-7d62-4cde-8680-e3e764228987.png)

**Comment by jumpjack** [comment]
Something strange happens: If I crop the image using IM saving to png, IM itself cannot find at all a subimage in the cropping:

`magick compare -metric RMSE -subimage-search frame-01-cropped-IM-PAINT.png subject.png  result.png`

```
compare: geometry does not contain image `frame-01-cropped-IM-PAINT.png' @ warning/transform.c/CropImage/594.
compare: images too dissimilar `frame-01-cropped-IM-PAINT.png' @ error/compare.c/CompareImagesCommand/1166.
```

But if I make the same cropping using Windows Paint, IM finds the subimage.

`magick compare -metric RMSE -subimage-search frame-01-cropped-PAINT.png subject.png  result.png`

`4891.48 (0.0746391) @ 30,32`

If I open in Paint the crop created by IM and save it again using Paint, IM can't find the subimage.

I think the forum engine will alter the images, anyway here they are:

Created by IM:
![frame-01-cropped-IM](https://user-images.githubusercontent.com/1620953/128070831-2474fb80-e9a5-4dc0-8c4a-87b948a2ee13.png)

Created by IM, saved by PAINT:
![frame-01-cropped-IM-PAINT](https://user-images.githubusercontent.com/1620953/128070708-d538be8c-149b-4dcd-809d-c3754d3183dc.png)

Created in PAINT:
![frame-01-cropped-PAINT](https://user-images.githubusercontent.com/1620953/128070716-3ac634be-4a46-4c39-b57d-ac52b4741f73.png)

Image to search:

![subject](https://user-images.githubusercontent.com/1620953/128071147-9b9c8ed7-a388-48b4-a73e-3c66c7f3008f.jpg)

**Reply by fmw42** [reply]
Show your crop command. As snibgo said, always add +repage after a crop. PNG carries the crop information internally otherwise and thinks the image is still full size. JPG does not carry that information, but I always add +repage so that I do not forget to remove the virtual canvas information (unless I really want it). TIFF also carries that information, I believe.

**Reply by fmw42** [reply]
Which image is the large one and which the small one?  The large one must come first. Also you probably need to do +repage on both images, if they were cropped from a  larger one.

**Comment by snibgo** [comment]
> Something strange happens: If I crop the image using IM saving to png,...

You haven't shown the IM command you used. Why not? Please ALWAYS show the command you used.

I guess that you didn't "+repage" after the crop. But that is just a guess.

**Reply by jumpjack** [reply]
Sorry, I used:

`magick frame-01.png -crop 100x80+160+130 frame-01-cropped-IM.png`

I have no idea of what "+repage" does and means.... The cropped image shows fine in Paint!
Anyway adding it solves (of course...) the issue, but only for RMSE; for NCC I still get tens of seconds of processing time.

```
magick frame-01.png -crop 100x80+160+130 +repage frame-01-cropped-IM.png
magick compare -metric RMSE -subimage-search frame-01-cropped-IM.png subject.png  result.png

```
One millisecond later...

```
4236.14 (0.0646394) @ 30,32
```
But:

```
magick compare -metric NCC -subimage-search frame-01-cropped-IM.png subject.png  result.png
```

25 boring seconds later:

```
0.96365 @ 30,32
```


Now I am struggling to figure out how to concatenate CROP and COMPARE in a single line in Windows, but no luck, I don't get it....

**Reply by fmw42** [reply]
As I said, only NCC is accelerated. And this was only for recent versions of ImageMagick. Anything else uses an older much slower method of processing. So best to use NCC when you can.

**Reply by jumpjack** [reply]
for unknown reasons, in my system it happens exatctly the opposite: NCC takes seconds, RMSE takes milliseconds.

**Reply by fmw42** [reply]
Do you have FFTW delegate installed?  What do you get from 

magick -version

if it does not list FFTW then you need to install that for Windows. You are probably testing only on a small pair of images. The accelerated NCC using FFTW will probably not show a lot of difference on small images. Was that on the full image above and the small subsection that I tested or on your two small cropped images?

**Reply by jumpjack** [reply]
No, FFTW does not appear in the list, only HDRI, but reading IM documentation I understood that it depends on how IM is compiled, not on what is installed on my system.
Or not?

I did the test both on full image and cropped one, it always takes 20-25 seconds, or it just fails, not finding anything.

**Reply by fmw42** [reply]
Install FFTW on your system. I do not know Windows, so cannot advise about how to do that. But see http://www.fftw.org/install/windows.html

**Reply by snibgo** [reply]
Pre-built distributions on Windows do not include fftw, so will not have the NCC acceleration.

The Windows source does include fftw, so should include the NCC acceleration, but I haven't tested this.

Windows source is at https://imagemagick.org/script/install-source.php#windows . Instructions there are for Visual Studio. Personally, I use the GNU toolset from Cygwin, which works fine.

**Comment by snibgo** [comment]
On using "+repage" after "-crop", see the documentation https://imagemagick.org/script/command-line-options.php#crop

On combining crop and compare into one command (Windows syntax):
```
magick compare ^
  frame-01.png -crop 100x80+160+130 +repage ^
  crop2.png ^
  -metric RMSE ^
  -subimage-search ^
  result.png
```

**Reply by jumpjack** [reply]
compare: unable to open image 'crop2.png': No such file or directory @ error/blob.c/OpenBlob/3537

**Reply by snibgo** [reply]
Use whatever filename you have used for that file.

**Reply by jumpjack** [reply]
Found the right method to concatenate crop and compare:

```
magick compare frame-01.png -crop 100x80+160+130 +repage   -metric RMSE -subimage-search subject.png result.png

4236.14 (0.0646394) @ 30,32
```

 - no need for "crop2.png" output: the "output" is  "frame-01.png -crop 100x80+160+130 +repage"
 - need for something to search in the cropped image: "subject.png"

**Comment by fmw42** [comment]
Best to use parentheses so that you could crop both images if desired.

On Windows

`magick compare  -metric RMSE -subimage-search ( large_image -crop W1xH1+X1+Y1 +repage ) ( small_image -crop W2XH2+X2+Y2 +repage ) result`

If on Unix-like systems, put \ before each paren as  \( ... \)

See https://legacy.imagemagick.org/Usage/basics/#parenthesis

**Comment by jumpjack** [comment]
Ok thanks, I managed it to work. Unfortunately I also discovered that stabilizing a video is not just a matter of detecting where a subimage of first frame moves in second frame, but a much more complex thing...
Fortunately I found out how to use the text output of deshaker filter of VirtualDub, so VirtualDub will do the job for me. :-) (At least, the first part, then I shall port its "flat stabilization" into "360 stabilization".

**Comment by snibgo** [comment]
I did video stabilization some years ago, for handheld DSLR footage, using ImageMagick. "-distort perspective" worked well. For that, we need four points in one frame and the corresponding four points in the other frames.

This works because a hand-held camera doesn't move much along the XYZ axes, compared to a typical subject distance. Instead, the camera rotates around those axes, creating noticeable movement between frames, and a perspective distortion compensates for this.

**Comment by jumpjack** [comment]
I am talking about stabilization of 360°x180° video projected into equirectangular format; standard stabilizers cannot work with this format.
